src/bridge/cockpitdisksamples.c

```diff
@@ -203,55 +202,49 @@ table_add_values(GHashTable *table,
 }
 
 static void
-get_process_io (const gchar *dir_name,
+get_process_io (const int dirfd,
                 GHashTable *table)
 {
-  g_autofree gchar *path = g_strdup_printf ("/proc/%s/io", dir_name);
-  g_autofree gchar *contents = read_file (path);
-  if (!contents)
+  FILE *io_fp = open_file (dirfd, "io");
+  if (!io_fp)
     return;
 
-  guint64 disk_read = 0, disk_write = 0;
-  gchar **lines = g_strsplit (contents, "\n", -1);
-  for (guint n = 0; lines != NULL && lines[n] != NULL; n++)
+  gchar *key;
+  guint64 disk_read = 0, disk_write = 0, value = 0;
+  while (fscanf (io_fp, "%m[^: ]: %lu\n", &key, &value) == 2)
     {
-      gchar description[32] = {'\0'};
-      guint64 value;
-      const gchar *line = lines[n];
-      if (strlen (line) == 0)
-        continue;
-
-      if (sscanf (line, "%s %" G_GUINT64_FORMAT, description, &value) != 2)
-        {
-          g_warning ("Error parsing line %d of file /proc/%s/io: `%s'", n, dir_name, line);
-          continue;
-        }
-
-      if (g_str_equal (description, "read_bytes:"))
+      if (g_str_equal (key, "read_bytes"))
         disk_read = value;
-      else if (g_str_equal (description, "write_bytes:"))
+      else if (g_str_equal (key, "write_bytes"))
         disk_write = value;
+
+      free (key);
     }
 
-  g_strfreev (lines);
+  fclose (io_fp);
 
   // get process cgroup
-  g_autofree gchar *cgroup_path = g_strdup_printf ("/proc/%s/cgroup", dir_name);
-  g_autofree gchar *cgroup = read_file (cgroup_path);
-  if (!cgroup)
+  FILE *cgroup_fp = open_file (dirfd, "cgroup");
+  if (!cgroup_fp)
     return;
-  g_strchomp (cgroup);
 
-  table_add_values (table, cgroup, disk_read, disk_write);
+  g_autofree gchar *cgroup = NULL;
+  if (fscanf (cgroup_fp, "%ms\n", &cgroup) != 1)
+    g_message ("Error when reading cgroup name: %m");
+  else
+    table_add_values (table, cgroup + 4, disk_read, disk_write);
+
+  fclose (cgroup_fp);
 }
 
 void
 cockpit_cgroup_disk_usage (CockpitSamples *samples)
 {
   g_autoptr(GError) error = NULL;
   g_autoptr(GDir) dir = g_dir_open ("/proc", 0, &error);
```

Comment:
I think I'd just trade out `GDir` for `opendir` and `readdir`, since then you get to use `dirfd()` and don't need to open the directory twice.

It also lets you check for the items being directories (although I think it's pretty pointless if we're dealing with procfs -- we already know all numeric entries are directories).

Reply:
Done, and I think it looks nicer